otaclient/app/common.py

```diff
@@ -306,83 +334,264 @@ def urljoin_ensure_base(base: str, url: str):
+    untill all the element in the input <_iter> is successfully processed,
+    or max retries exceeded the limitation.
+
+    NOTE: the inst of this class can only be used once.
+    NOTE: the API of this class is NOT thread-safe.
+
+    Args:
+        max_failed: <= 0 means no max_failed
+        max_retry: <=0 means no max_retry
+    """
+
+    def __init__(
         self,
-        extra_wait_cb: Optional[Callable] = None,
+        _func: Callable[[_T], Any],
+        _iter: Iterable[_T],
+        /,
         *,
-        raise_exception: bool = True,
-    ):
-        while (not self._register_finished) or self._done_num < self._in_num:
-            if self._interrupted.is_set():
-                logger.error(f"{self.title} interrupted, abort")
-                break
+        max_concurrent: int,
+        executor: Executor,
+        title: str = "",
+        backoff_max: int = 5,
+        backoff_factor: float = 1,
+        max_retry: int = 6,
+    ) -> None:
+        self.title = title
+        self.max_concurrent = max_concurrent
+        self._func = _func
+        self._executor = executor
+        self._max_retry_iter = range(max_retry) if max_retry > 0 else itertools.count()
+        # life cycle
+        self._status_lock = Lock()
+        self._status = _RetryTaskMapStatus.INIT
+
+        # tasks handling
+        self._collector_done_this_round = Event()
+        self._dispatcher_done_this_round = Event()
+        self._yield_que: Queue[Future] = Queue()
+        self._task_collector_gen: _ThreadSafeGenWrapper = None  # type: ignore
+        self._task_dispatcher_thread: Future = None  # type: ignore
+        self._last_error: Exception = None  # type: ignore
+
+        # values specific to each retry
+        self._iter: Iterable[_T] = _iter  # start from the input _iter
+        self._futs: Set[Future] = set()
+        self._failed_tasks: List[_T] = []
+        self._last_failed_count = 0
+        self._backoff_wait_f = partial(
+            wait_with_backoff, _backoff_factor=backoff_factor, _backoff_max=backoff_max
+        )
 
-            time.sleep(self._wait_interval)
+    def _task_collector(self) -> Generator[None, Future, None]:
+        while self._status is _RetryTaskMapStatus.RUNNING:
+            _fut = yield
+            try:
+                _exp, _entry, _ = _fut.result()
+                if _exp is not None:
+                    self._failed_tasks.append(_entry)
+                    self._last_error = _exp
+                # give the fut to the upper caller
+                self._yield_que.put_nowait(_fut)
+            except Exception:  # ignored as not caused by task itself
+                pass
+
+            # if the only element in self._futs is _ROUNT_START_SENTINEL,
+            # it means that this retry round is finished.
+            if (
+                self._dispatcher_done_this_round.is_set()
+                and len(self._futs) == 1
+                and _ROUND_START_SENTINEL in self._futs
```

Comment:
I don't know why this `_ROUND_START_SENTINEL` is added in the self._futs set and checked `if len(self._futs) == 1`.
Why remove `self._futs.add(_ROUND_START_SENTINEL)`, and `len(self._futs) == 0` is not enough?

Reply:
Actually this is a trick to let the collector generator awares of each retry round and prevent the collector from pre-maturely ending the collecting for this retry Round.
But yes the naming of this sentinel is kind of misleading, I will change it to `_RETRY_ROUND_SENTINEL`.

> len(self._futs) == 0 is not enough?

If the tasks finished faster than tasks being dispatched, there is a chance that the `self._futs` will be empty and new tasks are not yet dispatched and scheduled, causing the collector thinks it finishes its job in this retry round. I actually encountered the above situation when doing test, so I introduce this `_RETRY_ROUND_SENTINEL` mechanism to avoid pre-maturely round ending.

Reply:
> If the tasks finished faster than tasks being dispatched, there is a chance that the self._futs will be empty and new tasks are not yet dispatched and scheduled, causing the collector thinks it finishes its job in this retry round.

I think even if _ROUND_START_SENTINEL is introduced, the following condition will be satisfied on the above circumstance?
It means introducing `_ROUND_START_SENTINEL` can't solve this `finished faster than tasks being dispatched` condition.
```
and len(self._futs) == 1
and _ROUND_START_SENTINEL in self._futs
```

Reply:
sentinel here is not a `Future` inst created by the executor.sumbit function, so it will not be discard by done_callback, only the collector generator can remove it.

Reply:
`self._futs.add(_RETRY_ROUND_SENTINEL)` is done before any task submit.
So I think checking

```
and len(self._futs) == 1
and _ROUND_START_SENTINEL in self._futs
```

and 

```
and len(self._futs) == 0
```

are logically same if _RETRY_ROUND_SENTINEL is not set to self._futs.add.

Reply:
@obi-t4 Emmm you are right, let me reconsider again.

Reply:
I think we need to wait until length of self._iter and self._futs are the same in the done callback.

Reply:
@obi-t4 The `RetryTaskMap` is fixed in [67e43e8](https://github.com/tier4/ota-client/pull/193/commits/67e43e8a490671d82b2a53bab9ba63e45a1c84e0), please have a check, thank you!